src/support/thread_group.c

```diff
@@ -0,0 +1,330 @@
+		 * adjust it before finding the last worker in the group.
+		 */
+		worker = group->workers[--group->current_workers];
+		/* Be paranoid in case we are cleaning up after an error */
+		if (worker == NULL)
+			continue;
+
+		__wt_verbose(session, WT_VERB_UTIL_THREAD,
+		    "Stopping utility worker: %p:%"PRIu32"\n",
+		    group, worker->id);
+		F_CLR(worker, WT_WORKER_THREAD_RUN);
+		/* Wake threads to ensure they notice the state change */
+		__wt_cond_signal(session, group->wait_cond);
+		WT_TRET(__wt_thread_join(session, worker->tid));
+		worker->tid = 0;
+
+		/*
+		 * Worker thread sessions are only freed when shrinking the
+		 * pool or shutting down the connection.
+		 */
+		if (free_worker) {
+			if (worker->session != NULL) {
+				wt_session = (WT_SESSION *)worker->session;
+				WT_TRET(wt_session->close(wt_session, NULL));
+				worker->session = NULL;
+			}
+			__wt_free(session, worker);
+		}
+	}
+	return (ret);
+}
+
+/*
+ * __util_thread_group_resize --
+ *	Resize an array of utility workers already holding the lock.
+ */
+static int
+__util_thread_group_resize(
+    WT_SESSION_IMPL *session, WT_WORKER_THREAD_GROUP *group,
+    uint32_t new_min, uint32_t new_max, uint32_t flags)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_WORKER_THREAD *worker;
+	size_t alloc;
+	uint32_t i, session_flags;
+
+	conn = S2C(session);
+	session_flags = 0;
+
+	WT_ASSERT(session,
+	    group->current_workers <= group->alloc &&
+	    __wt_rwlock_islocked(session, group->lock));
+
+	if (new_min == group->min && new_max == group->max)
+		return (0);
+
+	if (group->current_workers > new_max)
+		WT_RET(__util_thread_group_shrink(
+		    session, group, new_max, true));
+
+	/*
+	 * Only reallocate the worker array if it is the largest ever, since
+	 * our realloc doesn't support shrinking the allocated size.
+	 */
+	if (group->alloc < new_max) {
+		alloc = group->alloc * sizeof(*group->workers);
+		WT_RET(__wt_realloc(session, &alloc,
+		    new_max * sizeof(*group->workers), &group->workers));
+		group->alloc = new_max;
+	}
+
+	/*
+	 * Initialize the structures based on the previous group size, not
+	 * the previous allocated size.
+	 */
+	for (i = group->max; i < new_max; i++) {
+		WT_ERR(__wt_calloc_one(session, &worker));
+		/*
+		 * Utility worker threads have their own session.
+		 *
+		 * Utility worker threads get their own lookaside table cursor
+		 * if the lookaside table is open.  Note that utility threads
+		 * are started during recovery, before the lookaside table is
+		 * created.
+		 */
+		if (LF_ISSET(WT_WORKER_CAN_WAIT))
+			session_flags = WT_SESSION_CAN_WAIT;
+		if (F_ISSET(conn, WT_CONN_LAS_OPEN))
+			FLD_SET(session_flags, WT_SESSION_LOOKASIDE_CURSOR);
+		WT_ERR(__wt_open_internal_session(conn, "utility-worker",
+		    false, session_flags, &worker->session));
+		if (LF_ISSET(WT_WORKER_PANIC_FAIL))
+			F_SET(worker, WT_WORKER_PANIC_FAIL);
+		worker->id = i;
+		worker->run_func = group->run_func;
```

Comment:
It appears the only reason to have a group function is to have it as a holding place for the worker structure to copy it.  It may be needed when adding a new thread unless the caller can pass it in.